src/Features/Core/Portable/CodeLens/ReferenceLocationDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.CodeLens
+{
+    /// <summary>
+    /// Holds information required to display and navigate to individual references
+    /// </summary>
+    internal sealed class ReferenceLocationDescriptor
+    {
+        public int LineNumber { get; }
+
+        public int ColumnNumber { get; }
+
+        public DocumentId DocumentId { get; }
+
+        /// <summary>
+        /// Language of the reference location
+        /// </summary>
+        public string Language { get; }
+
+        /// <summary>
+        /// Fully qualified name of the symbol containing the reference location
+        /// </summary>
+        public string LongDescription { get; }
+
+        /// <summary>
+        /// The kind of symbol containing the reference location (such as type, method, property, etc.)
+        /// </summary>
+        public Glyph? Glyph { get; }
+
+        /// <summary>
+        /// the full line of source that contained the reference
+        /// </summary>
+        public string ReferenceLineText { get; }
+
+        /// <summary>
+        /// the beginning of the span within reference text that was the use of the reference
+        /// </summary>
+        public int ReferenceStart { get; }
+
+        /// <summary>
+        /// the length of the span of the reference
+        /// </summary>
+        public int ReferenceLength { get; }
+
+        /// <summary>
+        /// Text above the line with the reference
+        /// </summary>
+        public string BeforeReferenceText1 { get; }
+
+        /// <summary>
+        /// Text above the line with the reference
+        /// </summary>
+        public string BeforeReferenceText2 { get; }
+
+        /// <summary>
+        /// Text below the line with the reference
+        /// </summary>
+        public string AfterReferenceText1 { get; }
+
+        /// <summary>
+        /// Text below the line with the reference
+        /// </summary>
+        public string AfterReferenceText2 { get; }
```

Comment:
In the future we should be able to make this much nicer.  For example, see the work done with the new streaming FAR system.  There we actually are able to show a real preview of the actual code of the reference, not just unclassified lines of text.  But that can wait.